src/pushsource/_impl/backend/staged/staged_source.py

```diff
@@ -136,6 +137,25 @@ def _push_items_for_leafdir(self, leafdir, metadata):
     def _push_items_for_topdir(self, topdir):
         LOG.info("Checking files in: %s", topdir)
 
+        timeout = int(os.getenv("PUSHSOURCE_SRC_POLL_TIMEOUT") or "0")
+        poll_rate = int(os.getenv("PUSHSOURCE_SRC_POLL_RATE") or "30")
+
+        # If staged path doesn't exists, we can assume that once it does, it will
+        # contain fresh content. It's sufficient to only wait until it appears.
+        if not os.path.exists(topdir):
+            wait_exist(topdir, timeout, poll_rate)
+        # If the path already exists, it may have been modified which means that pushsource
```

Comment:
I'm curious whether this has been inspired by any real problems encountered yet, or is it only anticipating problems?

I'm not sure that we should add this.  I think the users should be strongly discouraged from modifying staging dirs after creation, and they should instead create new ones (or at least a new *path*, e.g. by renaming the dir).

Mutable staging dirs would always have caused some problems even before recent Pub changes:

- the filesystem was always synced to multiple regions with some delay. So by modifying the staging area after creation, they've always been creating multiple inconsistent views of the staging area.
- if staging area is modified after initial use, it becomes impossible to reproduce what Pub did in a given push, and difficult to debug if something unexpected happened.

Basically, if path X is allowed to contain different content at different times, and there is caching or replication in play (and there always has been), then life is harder for everyone.  So the users shouldn't be doing that.

It seems like this change is going to impose a severe penalty on almost *all* users of push-staged in order to make life easier for those users doing something they shouldn't be doing it in the first place.

It doesn't seem right to me.  How about just telling the users/letting them learn that, if they need to change a staging area after creation, it's better to create a new one, rename the dir, or wait 15 minutes?  This means the users who are modifying their staging areas will feel the pain and be encouraged to change their processes, rather than *all* the users feeling pain even if they're using the system in an ideal way.

Reply:
I think in the past it often happen there was typo in repo directory or missing product ids. I'm not sure how often it was fixed in place or how often new directory was created. But I assume relengs usually choose the easiest way.